### PR TITLE
Granting wish #12762: warning on duplicated catch-all pattern-matching clause with unused named variable

### LINKED ISSUE #12762: [request] Add a warning for non-underscore catchall cases
Suppose I have the code
```coq
Inductive foo := a | b.

Definition bar (f : foo) :=
  match f with
  | a => 0
  | B => 1
  end.
```
This is fine, but if I extend the inductive with `| c`, then Coq won't tell me that I need to extend the `match`.  I propose that a warning is emitted in cases where there is a named catch-all case for a `match` where the variable is unused in the branch (so, e.g., `match f with a => b | B => B end` wouldn't trigger the warning).

### PR DESCRIPTION
**Pull request by herbelin**
**Kind:** enhancement

Fixes / closes #12762

No strong idea on how the warning should be named...

- [X] Added / updated test-suite
- [X] Entry added in the changelog

### CODE REVIEW

**Comment by gares** [comment]
@Zimmi48 would it be possible to have coqbot not report about allowed failure jobs when build:base fails? When I see the Coqbot message about async I tend to click on it, ask myself "how it is possible that this is an async problem?" and then discover that Coq just does not build. Maybe this is not the right place to report the problem, but it is a trivial instance of the problem.

**Comment by Zimmi48** [comment]
@gares This indeed doesn't sound like the right place to discuss this. Do you mind moving this discussion to https://coq.zulipchat.com/#narrow/stream/243318-coqbot-devs.20.26.20users?

**Comment by herbelin** [comment]
Here are excerpts of a couple of situations where the stdlib gives a name to a catch-all clause:
```
match m with                
    | Leaf => Leaf       
    | Node Leaf o Leaf => Leaf              
    | Node l o r => Node l None r 
```
```
match t1 , t2 with                                                                                              
    | PsatzZ      , x        => PsatzZ                                                  
    |    x     , PsatzZ      => PsatzZ                                                         
    | PsatzC c ,  PsatzC c' => PsatzC (ctimes c c')
```
```
  match n with                                                                       
  | 0 => fun _ => []                                    
  | S n' =>                                                  
    fun v =>                                                
      let (b0, v1) := uncons v in                   
      let (b1, v2) := uncons v1 in                  
      let (b2, v3) := uncons v2 in                
      let (b3, v4) := uncons v3 in                
      let (b4, v5) := uncons v4 in                
      let (b5, v6) := uncons v5 in                
      let (b6, v7) := uncons v6 in                
      let (b7, v8) := uncons v7 in                 
      Ascii b0 b1 b2 b3 b4 b5 b6 b7::of_Bvector v8 
```
```
  Fixpoint compare x y :=                                                                                             
    match x, y with                                                                                                   
      | x~1, y~1 => compare x y                                                                                       
      | x~1, _ => Gt
```
This does not seem illegitimate uses. Should we restrict the warning to toplevel cases (i.e. not in the `node l o r`, `S n'` or `x~1` case)? Should we adopt the OCaml convention to have a `_` in front of an unused variable to avoid the warning?

**Comment by gares** [comment]
> This does not seem illegitimate uses. Should we restrict the warning to toplevel cases (i.e. not in the `node l o r`, `S n'` or `x~1` case)? Should we adopt the OCaml convention to have a `_` in front of an unused variable to avoid the warning?

+1 for the OCaml convention, but here things are a bit different IMO.

The issue reports a pattern like `| X =>` which is not only fragile but, to me eyes, also wrong because in CIC what you can name are the subterms, not the entire term. To me `| SomeNameNotAConstructor =>` is an error, not just a warning. IMO a catch-all pattern should be `_`, and only `_`.

Detecting unused, and not `_prefixed` variables, is a nice warning IMO, but does not seem exactly what the issue is about.

**Comment by gares** [comment]
At the same time a pattern like `| Node l o r =>` that does not use `o` is OK. One could warn, but at least `o` is a legit subterm.

**Comment by jashug** [comment]
@gares 
Out of curiosity, what would you say about code like
```coq
Definition break_ties (x y : nat)
  := match Nat.compare x y with
     | Eq => Gt
     | other => other
     end.
```
(To be clear, this is an example I just made up not code I found in the wild)

**Comment by gares** [comment]
Hum I see, I guess I rushed to conclusions.

Still I've sen OCaml code like `| (Gt|Lt) as other => other`. I do It myself when the data type is not written in stone like this one, eg it may get new constructors.

Maybe this discipline is enough to cover Jason's use case?

**Comment by JasonGross** [comment]
Yeah, I'm fine with forbidding named catch-all cases all-together as long as I can do `| _ as other => other`.  (It's very important to not have to list all catch-all cases when I want to name them, as there can be hundreds of cases if my other branches are deep pattern matches.). And I'm fine with only getting the warning for top-level non-constructors, though I do think out might sometimes be useful to have the warning for non-top-level unused names (I'd also be fine with this warning being off-by-default.)

**Comment by gares** [comment]
> Yeah, I'm fine with forbidding named catch-all cases all-together as long as I can do `| _ as other => other`.

I was thinking the same. But I'm afraid there is a lot of code out there doing `| other => other` :-/

**Comment by JasonGross** [comment]
If it's made an error that can be turned into a warning, it should be pretty easy to write a python script or something that will process the log when it's set to a warning and replace `|` with `| _ as` in all the locations where the warning shows up.  It seems to me that having a fully automated way of generating backwards-compatible patches should be enough, even if the pattern shows up often in the wild.  Ideally we'd get a sightly different error message when the name is used vs unused, so that we could automatically handle both `| other => other` and `| other => something` separately in the script

**Comment by JasonGross** [comment]
Based on https://github.com/coq/coq/issues/12762#issuecomment-667700218, it looks like we do, in fact, want a version of the warning that includes unused subterm variables which are not constructors, and we want to have this warning enabled in the stdlib (as it would have prevented an inconsistency introduced into the stdlib).

**Comment by herbelin** [comment]
What are the conclusions:
- A warning on all unused (non-constructor) variables, either at toplevel or in subpatterns, except if the variable starts with `_` or is given with the form `(_ as foo)`?
- A warning even if the variable is used (i.e. `| other => pat(other)` to be turned into `| _ as other => pat(other)`)? That would be a rather significant change of habits but why not.
- Different warnings depending on whether it is in a strict subpattern position or not?

Note: Distinguishing `_ as foo` from `foo` would be slightly more intrusive in the implementation (because the information is in constrintern but lost in pretyping), but that's possible. Distinguishing the toplevel and a subpattern would also require to pass around a bit more information.

**Comment by gares** [comment]
I vote for one, but I don't know this code is so I can't comment on the extra complexity it requires.

**Review by vbgl — APPROVED** [review]
The changes to the Coq libraries look good.

**Comment by herbelin** [comment]
Report about the errors:
- In CoLoR, CompCert, MetaCoq and VST, the internally-generated `Program` catch-all variable `wildcard` is explicitly referred to by the user. But for the purpose of the PR, I renamed it to `_wildcard`. A decision has to be made whether we keep the name `wildcard` (but find another way to avoid the warning, e.g. by implementing support for `_ as wildcard`) or if we change CoLoR so that it uses `_wildcard` instead (and maybe document the `wildcard` name?).
- In AAC tactics, warnings are on by default, waiting for coq-community/aac-tactics#65 to be merged.

**Comment by herbelin** [comment]
I followed a lazy path, backtracked on renaming the internal `Program` name `wildcard'` into `_wildcard'` and instead do so that this name does not generate warnings (since several developments rely explicitly on it). I'm renouncing to support `_ as id` as a way to not generate the warning (this would require changing `Glob_term.cases_pattern`) though the idea remains good per se.

**Comment by herbelin** [comment]
If you're ok with the compromise (i.e. renouncing to support `_ as x` as a way to avoid the warning), this is ready. @gares, would you assign?

File: doc/changelog/02-specification-language/12768-master+warn-non-underscore-catch-all-pattern-matching.rst
Comment (gares) [inline doc/changelog/02-specification-language/12768-master+warn-non-underscore-catch-all-pattern-matching.rst:2]: I'm not in love with this wording, to me a catch-all branch is (at least) "duplicate".

IMO you should just say that we warn on unused variables in pattern matching branches.

Then, in the refman, describe as precisely as possible the current condition and convention (about `_` for example).

File: pretyping/cases.ml
Comment (gares) [inline pretyping/cases.ml:132]: It's a nit, but I believe you should use `mutable` here rather than `ref` (if you do so you don't need the `!` to deref)

File: doc/changelog/02-specification-language/12768-master+warn-non-underscore-catch-all-pattern-matching.rst
Comment (herbelin) [reply doc/changelog/02-specification-language/12768-master+warn-non-underscore-catch-all-pattern-matching.rst:2]: I agree that this is not fully clear. I made a new attempt. See also the comment about duplication below.

I add a word in the refman. Tell me if you find it clear enough (up to the duplicate question).

**Comment by herbelin** [comment]
I realize that I misread the suggestion at #12762. I activated the warning only when the variable appears in a catch-all which is used to complete at least two distinct patterns of the pattern-matching problem.

If I activate it even in non-duplicated catch-all clauses, the warnings is triggered 124 more times in the stdlib (and probably many many times elsewhere), which starts to make a lot.

File: doc/changelog/02-specification-language/12768-master+warn-non-underscore-catch-all-pattern-matching.rst
Comment (jfehrle) [inline doc/changelog/02-specification-language/12768-master+warn-non-underscore-catch-all-pattern-matching.rst:4]:
```suggestion
  Warning on unused variables in pattern-matching branches of
  :n:`match` serving as catch-all branches for at least two distinct
  patterns.
```

File: pretyping/cases.ml
Comment (jfehrle) [inline pretyping/cases.ml:555]:
```suggestion
          strbrk "Unused variable " ++ Id.print id ++ strbrk " catches more than one case.")
```

File: doc/sphinx/language/extensions/match.rst
Comment (jfehrle) [inline doc/sphinx/language/extensions/match.rst:302]:
```suggestion
Convention about wildcard patterns
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

By default, there is a warning when a pattern variable matches
two distinct patterns and is not used
on the right-hand side of ``=>``. The warning can be deactivated by using a
variable name starting with ``_`` or by setting ``Set Warnings
"-duplicated-catch-all-with-unused-named-variable"``.

Here is an example where the warning is activated.
```

Ah, several thoughts:
- Can we find a better title?  I have no idea from the title what the paragraph is about.
- Would you put the message in a `.. warn::` here so it appears in the index?  Then put some of the explanation below it?
- Can we find a shorter name for `duplicated-catch-all-with-unused-named-variable`?

**Review by jfehrle — COMMENTED** [review]
Doc suggestions.  Didn't look at the code.

**Comment by gares** [comment]
@herbelin the changes proposed by Jim look good to me, also there is a compilation problem in the refman.
After that I think we are good to merge.

File: doc/sphinx/language/extensions/match.rst
Comment (herbelin) [reply doc/sphinx/language/extensions/match.rst]: > - Can we find a better title?  I have no idea from the title what the paragraph is about.

What about "Convention about unused pattern-matching variables".

> - Would you put the message in a `.. warn::` here so it appears in the index?  Then put some of the explanation below it?

OK, I added a `.. warn::`` and reformulated the paragraph. A second reading might be useful.

> - Can we find a shorter name for `duplicated-catch-all-with-unused-named-variable`?

We should... "duplicated-unused-pattern-matching-variable" is still long. Otherwise, "unused-pattern-matching-variable"?

File: doc/sphinx/language/extensions/match.rst
Comment (jfehrle) [inline doc/sphinx/language/extensions/match.rst:297]:
```suggestion
Pattern-matching variables that are not used on the right-hand side of ``=>`` are
considered the sign of a potential error. For instance, it could
```

File: doc/sphinx/language/extensions/match.rst
Comment (jfehrle) [reply doc/sphinx/language/extensions/match.rst]: How about "Reporting unused pattern-matching variables"?

**Review by jfehrle — APPROVED** [review]
LGTM.

File: doc/sphinx/language/extensions/match.rst
Comment (herbelin) [reply doc/sphinx/language/extensions/match.rst]: Thanks! (And sorry for the "is".)

**Comment by herbelin** [comment]
I believe that this is ready.

**Comment by gares** [comment]
@coqbot merge now please

**Pull request merged**
